ci-automation/garbage_collect_releases.sh

```diff
@@ -0,0 +1,183 @@
+    (
+        set -euo pipefail
+
+        _garbage_collect_releases_impl "${@}"
+    )
+}
+# --
+
+function _garbage_collect_releases_impl() {
+    local keep_per_chan="${1:-10}"
+    local keep_lts_releases="${2:-2}"
+    local dry_run="${DRY_RUN:-}"
+
+    echo
+    echo "Number of versions to keep per channel: '${keep_per_chan}'"
+    echo "Number of LTS major releases to keep: '${keep_lts_releases}'"
+    echo
+
+    source ci-automation/ci_automation_common.sh
+    local sshcmd="$(gen_sshcmd)"
+
+    local keep="$( 
+        # For some reasons this is set to the empty string in some environments and it makes gawk
+        #  behave like POSIX awk (i.e. no 'gensub').
+        unset POSIXLY_CORRECT
+        curl -s "${RELEASES_JSON_FEED}" \
+            | jq -r 'keys_unsorted | .[] | match("[0-9]+\\.[0-9]+\\.[0-9]+") | .string' \
+            | sort -Vr \
+            | awk -v keep="${keep_per_chan}" -v lts="${keep_lts_releases}" '
+                {
+                    version = $1
+                    chan_num = gensub("[0-9]+\\.([0-9]+)\\.[0-9]+","\\1","g", version) + 0
+                    major = gensub("([0-9]+)\\.[0-9]+\\.[0-9]+","\\1","g", version) + 0
+
+                    if (chan_num <= 2) {
+                        if (chan_count[chan_num] < keep)
+                            print version
+                        chan_count[chan_num] = chan_count[chan_num] + 1
+                    } else {
+                        if (    (chan_count["lts"][major] < keep) \
+                             && (length(chan_count["lts"]) <= lts) )
+                            print version
+                        chan_count["lts"][major] = chan_count["lts"][major] + 1
+                    }
+                } '
+    )"
+
+    mapfile -t keep_versions <<<"${keep}"
+
+    echo
+    echo "######## The following version(s) will be kept ########"
+    if [ "$dry_run" = "y" ] ; then
+        echo
+        echo "(NOTE this is just a dry run since DRY_RUN=y)"
+        echo
+    fi
+    printf "%s\n" "${keep_versions[@]}"
+
+    local dir=""
+    for dir in  "sdk/amd64" \
+                "containers" \
+                "boards/amd64-usr" \
+                "boards/arm64-usr" \
+                "images/amd64" \
+                "images/arm64" \
+                "testing" \
+                ; do
+
+        local fullpath="${BUILDCACHE_PATH_PREFIX}/${dir}"
+        echo
+        echo "## Processing '${fullpath}'"
+        echo "---------------------------"
+        for version in $($sshcmd "${BUILDCACHE_USER}@${BUILDCACHE_SERVER}" \
+                                 "ls -1 ${BUILDCACHE_PATH_PREFIX}/${dir} | grep -E '^[0-9]+\.[0-9]+\.[0-9]+$'"); do
+            local o_fullpath="${fullpath}/${version}"
+
+            # skip if version is marked for keeping OR if it's a new release about to be published
+            if printf "%s\n" "${keep_versions[@]}" \
+               | { unset POSIXLY_CORRECT ; awk -v version="${version}" -v path="${dir}" '
```

Comment:
Can I ask you to rename some of the variables? It would make it easier to track which version, major, minor and patch is which. What I mean is something like renaming `version` to `bincache_version`, `v{major,minor,patch}` to `bincache_{major,minor,patch}`. And then possibly having an assignment like `keep_version = $0` and variables `keep_{major,minor,patch}`. `major_alpha` probably could be `keep_major_alpha`.

Reply:
Yes, good point. Tbh I got confused myself when updating the PR after one week of vacation :sweat_smile: